internal/db/admin.go

```diff
@@ -0,0 +1,668 @@
+		permissionNeeded != acpTypes.AdminAACStatusPerm {
+		return ErrAACIsNotConfigured
+	}
+
+	ident := identity.FromContext(ctx)
+	if ident.HasValue() &&
+		db.nodeIdentity.HasValue() &&
+		ident.Value().DID() == db.nodeIdentity.Value().DID() {
+		return nil
+	}
+
+	return CheckAACNodeOperationAccess(
+		ctx,
+		ident,
+		db.adminInfo,
+		permissionNeeded,
+		acpTypes.InternalAdminACPObjectToGate,
+	)
+}
+
+// CheckAACNodeOperationAccess returns an [client.ErrNotAuthorizedToPerformOperation]
+// error if the requesting user does not have the required permission to perform an operation.
+// If something else goes wrong, it returns a different error, otherwise returns nil only if
+// the check passes and the requesting user is authorized to perform the operation.
+//
+// Unrestricted access if:
+// - admin acp system is temporarily disabled (unless the operation is trying to turn it on aac).
+func CheckAACNodeOperationAccess(
+	ctx context.Context,
+	identity immutable.Option[acpIdentity.Identity],
+	adminInfo AdminInfo,
+	permission acpTypes.ResourceInterfacePermission,
+	objectID string,
+) error {
+	if !adminInfo.AdminDesc.IsEnabled && permission != acpTypes.AdminAACReEnablePerm {
+		// Unrestricted access if admin acp is off, and not trying to turn it back on.
+		return nil
+	}
+
+	// If admin acp is enabled then it must have have admin acp instance setup.
+	if adminInfo.AdminACP == nil {
+		return ErrAACIsEnabledButInstanceIsNotAvailable
+	}
+
+	// If admin acp is enabled then it must have a valid policy information.
+	if !adminInfo.AdminDesc.Policy.HasValue() {
+		return ErrAACIsEnabledButIsMissingPolicyInfo
+	}
+
+	policyID := adminInfo.AdminDesc.Policy.Value().ID
+	resourceName := adminInfo.AdminDesc.Policy.Value().ResourceName
+	if policyID == "" || resourceName == "" {
+		return ErrAACIsEnabledButIsMissingPolicyInfo
+	}
+	// Since public node will have unrestricted access, the object we are gating MUST be registered
+	// if admin access control is configured.
+	isRegistered, err := adminInfo.AdminACP.ObjectOwner(
+		ctx,
+		policyID,
+		resourceName,
+		objectID,
+	)
+	if err != nil {
+		return err
+	}
+
+	if !isRegistered.HasValue() {
+		return ErrAACNodeObjectToGateIsNotRegistered
+	}
+
+	var identityValue string
+	if !identity.HasValue() {
+		// We can't assume that there is no-access just because there is no identity even if the operation
+		// is registered with acp, this is because it is possible that acp has a registered relation targeting
+		// "*" (any) actor which would mean that even a request without an identity might be able to access
+		// an operation registered with acp. So we pass an empty `did` to accommodate that case.
+		identityValue = ""
+	} else {
+		identityValue = identity.Value().DID()
+	}
+
+	adminResourcePerm, ok := permission.(acpTypes.AdminResourcePermission)
+	if !ok {
+		return client.ErrInvalidResourcePermissionType
+	}
+
+	// Now actually check using the signature if this identity has access or not.
```

Comment:
question: What signature are we talking about? Following the call path there is no mention of signature so it's a bit confusing to see this here.

Reply:
Ah, this was my bad. I don't know why that was there, thanks for pointing out.

- [x] Fix comment